qiskit/extensions/quantum_initializer/state_preparation.py

```diff
@@ -0,0 +1,462 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2017.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+"""
+Prepare desired state
+"""
+import math
+import numpy as np
+
+from qiskit.exceptions import QiskitError
+from qiskit.circuit import Instruction, QuantumCircuit, QuantumRegister
+from qiskit.circuit.library.standard_gates.x import CXGate, XGate
+from qiskit.circuit.library.standard_gates.h import HGate
+from qiskit.circuit.library.standard_gates.s import SGate, SdgGate
+from qiskit.circuit.library.standard_gates.ry import RYGate
+from qiskit.circuit.library.standard_gates.rz import RZGate
+from qiskit.circuit.exceptions import CircuitError
+
+_EPS = 1e-10  # global variable used to chop very small numbers to zero
+
+
+class StatePreparation(Instruction):
+    """Complex amplitude state preparation.
+
+    Class that implements the (complex amplitude) state preparation of some
+    flexible collection of qubit registers.
+    """
+
+    def __init__(self, params, num_qubits=None):
+        """Prepare state
```

Comment:
This class should at least explain that it's only correct state prep from an initial state of `|00...0>`, but it might even be better (and true to the original issue) if it accepted two possible arguments: `start_state` and `end_state`, where `start_state` defaults to `|00...0>`.